### PR TITLE
Enable modular logging

### PR DESCRIPTION
**Pull request by sadiq1971**

### CODE REVIEW

File: packages/lodestar/src/cli/commands/beacon.ts [hunk 1]
```diff
@@ -50,10 +52,18 @@ export class BeaconNodeCommand implements CliCommand {
 
   public async action(options: IBeaconCommandOptions, logger: ILogger): Promise<void> {
     let conf: Partial<IBeaconNodeOptions> = {};
+    let loggingOptions;
```
Comment (GregTheGreek) [inline packages/lodestar/src/cli/commands/beacon.ts]: Can you set a type/interface for this?

File: packages/lodestar/src/cli/commands/eth1-private-network.ts [hunk 2]
```diff
@@ -45,10 +47,23 @@ export class Eth1PrivateNetworkCommand implements CliCommand {
       });
   }
 
-  public async action(options: IEth1CommandOptions, logger: ILogger): Promise<PrivateEth1Network> {
+  public async action(options: IEth1CommandOptions, logger): Promise<PrivateEth1Network> {
+    let loggingOptions: ILoggingOptions;
     if (options.loggingLevel) {
-      logger.setLogLevel(LogLevel[options.loggingLevel]);
+      loggingOptions = {
+        loggingLevel: LogLevel[options.loggingLevel],
```
Comment (GregTheGreek) [inline packages/lodestar/src/cli/commands/eth1-private-network.ts]: you can collapse the if statement into one

`loggingLevel: LogLevel[options.loggingLevel] || LogLevel.DEFAULT,`

Comment (GregTheGreek) [reply packages/lodestar/src/cli/commands/eth1-private-network.ts]: this applies to all similar instances

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

**Review by mpetrunic — CHANGES_REQUESTED** [review]
This just created a bunch of duplicated code.
Few points to correct:
- logger constructor should handle setting log level and it's default value so you don't need if statements everywhere
- each module (not cli command) should create new logger instance and take logging level that belongs to it
- cli option should look like this "eth1=debug,networking=trace" that means you will have to parse it into required format (I would suggest map module => level) and pass that map to every module as an option

File: packages/lodestar/src/chain/chain.ts [hunk 3]
```diff
@@ -33,9 +24,11 @@ import {getAttestingIndices, slotToEpoch} from "./stateTransition/util";
 import {IBeaconChain} from "./interface";
 import {ProgressiveMerkleTree} from "../util/merkleTree/merkleTree";
 import {processSortedDeposits} from "../util/deposits";
+import {Module} from "../logger/abstract";
+import {IChainOptions} from "./options";
 
 export class BeaconChain extends EventEmitter implements IBeaconChain {
-  public chain: string;
+  public chain: IChainOptions;
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: remove

File: packages/lodestar/src/chain/chain.ts [hunk 4]
```diff
@@ -46,13 +39,14 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
   private _latestBlock: BeaconBlock;
   private logger: ILogger;
 
-  public constructor(opts, {config, db, eth1, logger}: { config: IBeaconConfig; db: IBeaconDb; eth1: IEth1Notifier; logger: ILogger }) {
+  public constructor(opts, {config, db, eth1, logger}:
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: add IChainOptions type to opts

File: packages/lodestar/src/chain/options.ts [hunk 5]
```diff
@@ -1,8 +1,11 @@
 import {IConfigurationModule} from "../util/config";
 import {number64} from "../types";
+import {ILoggingOptions} from "../logger/option";
+import {LoggingOptions} from "../logger/option";
 
 export interface IChainOptions {
   name: string;
+  loggingOptions?: ILoggingOptions;
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/options.ts]: no need for object, it can just be `loggingLevel`: LoggingLevel

File: packages/lodestar/src/chain/options.ts [hunk 6]
```diff
@@ -21,7 +24,8 @@ export const ChainOptions: IConfigurationModule = {
         flag: "chain",
         short: "c"
       }
-    }
+    },
+    LoggingOptions
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/options.ts]: remove

File: packages/lodestar/src/cli/commands/beacon.ts [hunk 7]
```diff
@@ -35,26 +35,22 @@ export class BeaconNodeCommand implements CliCommand {
       .command("beacon")
       .description("Start lodestar node")
       .option("-c, --configFile [config_file]", "Config file path")
-      .option(`-l, --loggingLevel [${Object.values(LogLevel).join("|")}]`, "Logging level")
+      .option("-l, --loggingLevel [chain=debug, network=trace, database=warn]",
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/beacon.ts]: remove , those things have to be generated trough ConfigurationModule

Comment (mpetrunic) [reply packages/lodestar/src/cli/commands/beacon.ts]: remove loggingLevel from IBeaconCommandOptions

File: packages/lodestar/src/cli/commands/beacon.ts [hunk 8]
```diff
@@ -65,7 +61,7 @@ export class BeaconNodeCommand implements CliCommand {
     //override current config with cli config
     conf = deepmerge(conf, optionsToConfig(options, BeaconNodeOptions));
 
-    this.node = new BeaconNode(conf, {config, logger});
+    this.node = new BeaconNode(conf, {config,logger});
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/beacon.ts]: add whitespace back

File: packages/lodestar/src/cli/commands/beacon.ts [hunk 9]
```diff
@@ -85,14 +81,17 @@ export class BeaconNodeCommand implements CliCommand {
           {config, chain: this.node.chain, db: this.node.db}
         ),
       });
+
       this.validator = new Validator(
-        conf.validator,
-        {config, logger}
+        conf.validator, {
+          config: config,
+          logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/beacon.ts]: let validator create it's own logger with level for Module.Validator

File: packages/lodestar/src/cli/commands/eth1-private-network.ts [hunk 10]
```diff
@@ -28,7 +28,8 @@ export class Eth1PrivateNetworkCommand implements CliCommand {
       .description('Start private eth1 chain with deposit contract and 10 accounts with balance')
       .option("-p, --port [port]", 'Port on which private network node should start', 8545)
       .option("-h, --host [host]", 'Host on which node will be', '127.0.0.1')
-      .option(`-l, --loggingLevel [${Object.values(LogLevel).join("|")}]`, "Logging level")
+      .option("-l, --loggingLevel [chain=debug, network=trace, database=warn]",
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/eth1-private-network.ts]: there is no point in modular logging here

Comment (mpetrunic) [reply packages/lodestar/src/cli/commands/eth1-private-network.ts]: just let them pick logging level

File: packages/lodestar/src/cli/commands/eth1-private-network.ts [hunk 11]
```diff
@@ -37,28 +38,26 @@ export class Eth1PrivateNetworkCommand implements CliCommand {
...
+  public async action(options: IEth1CommandOptions, logger?: ILogger): Promise<PrivateEth1Network> {
     const privateNetwork = new PrivateEth1Network({
       port: options.port,
       host: options.host,
       mnemonic: options.mnemonic,
       networkId: options.network,
-      dbPath: options.database
+      dbPath: options.database,
+      loggingOptions: {
+        loggingLevel: parseLoggingLevel(options.loggingLevel),
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/eth1-private-network.ts]: whats the point of logginOptions in that class

File: packages/lodestar/src/cli/commands/validator.ts [hunk 12]
```diff
@@ -19,31 +19,34 @@ export class ValidatorCommand implements CliCommand {
 
   public register(commander: CommanderStatic): void {
     const logger: ILogger = new WinstonLogger();
-
     const command = commander
       .command("validator")
       .description("Start lodestar validator")
-      .option(`-l, --loggingLevel [${Object.values(LogLevel).join("|")}]`, "Logging level")
+      .option("-l, --loggingLevel [chain=debug, network=trace, database=warn]",
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/validator.ts]: there is no modules here, just let users pick logging level for validator module and pass it on as option

File: packages/lodestar/src/logger/option.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+/**
+ * @module logger
+ */
+
+import {LogLevel, Module} from "./abstract";
+import {IConfigurationModule} from "../util/config";
+import {parseLoggingLevel} from "../util/parse";
+
+export interface ILoggingOptions {
+  loggingLevel: Map<Module, LogLevel>;
```
Comment (mpetrunic) [inline packages/lodestar/src/logger/option.ts]: should be just LogLevel, map configuration should be on BeaconNodeOptions

**Review by mpetrunic — CHANGES_REQUESTED** [review]
To make it short and explain better:
- add loggingOptions ot IBeaconNodeOptions
- add logginOptions cli configuration to BeaconNodeOptions
- beacon node should in constructor :
```
this.opPool = new OpPool(
{
    {
        ...this.conf.opPool,
        logginLevel: this.config.logginOptions[Module.OpPool]
    },
   {chain, db, whatever}
}
);
```
- each module should:
```
public constructor(opts, {}) {
   opts = deepmerge(defaultOpts, opts);
this.logger = new WinstonLogger(Module.modulename, opts.loggingLevel)
}
```
so each module options should replace loggingOptiosn with logging level without cli configuration as cli configuration is on BeaconNode level.

Other commands should not go trought that process, they can work like they did till now, just have logginLevel in cli. You can also make module optional in WinstonLogger

File: packages/lodestar/src/chain/chain.ts [hunk 14]
```diff
@@ -45,14 +39,17 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
   private eth1: IEth1Notifier;
   private _latestBlock: BeaconBlock;
   private logger: ILogger;
+  private opts: IChainOptions;
 
-  public constructor(opts, {config, db, eth1, logger}: { config: IBeaconConfig; db: IBeaconDb; eth1: IEth1Notifier; logger: ILogger }) {
+  public constructor(opts: IChainOptions, {config, db, eth1, logger}:
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: Partial<IChainOptions>

File: packages/lodestar/src/cli/commands/eth1-private-network.ts [hunk 15]
```diff
@@ -37,28 +37,24 @@ export class Eth1PrivateNetworkCommand implements CliCommand {
...
-    }
+  public async action(options: IEth1CommandOptions, logger?: ILogger): Promise<PrivateEth1Network> {
     const privateNetwork = new PrivateEth1Network({
       port: options.port,
       host: options.host,
       mnemonic: options.mnemonic,
       networkId: options.network,
-      dbPath: options.database
+      dbPath: options.database,
+      loggingLevel: options.loggingLevel as LogLevel
```
Comment (mpetrunic) [inline packages/lodestar/src/cli/commands/eth1-private-network.ts]: does this work?

File: packages/lodestar/src/logger/winston.ts [hunk 16]
```diff
@@ -2,29 +2,38 @@
...
 
   private winston: Logger;
+  private loggingModule;
+  private loggingLevel;
 
-  public constructor() {
+  public constructor(loggingLevel?: LogLevel, loggingModule?: Module ) {
     super();
+    this.loggingModule = loggingModule || Module.DEFAULT;
+    this.loggingLevel = loggingLevel || LogLevel.DEFAULT;
```
Comment (mpetrunic) [inline packages/lodestar/src/logger/winston.ts]: you can assign that in constructor signature

File: packages/lodestar/src/network/libp2p/network.ts [hunk 17]
```diff
@@ -31,18 +32,19 @@ export class Libp2pNetwork extends EventEmitter implements INetwork {
   private inited: Promise<void>;
   private logger: ILogger;
 
-  public constructor(opts: INetworkOptions, {config, libp2p,logger}: {config: IBeaconConfig; libp2p: any; logger: ILogger}) {
+  public constructor(opts: INetworkOptions, {config, libp2p, logger}:
```
Comment (mpetrunic) [inline packages/lodestar/src/network/libp2p/network.ts]: `Partial<INetworkOptions>`

File: packages/lodestar/src/network/libp2p/syncRpc.ts [hunk 18]
```diff
@@ -8,52 +8,63 @@ import {hashTreeRoot} from "@chainsafe/ssz";
...
   private db: IBeaconDb;
   private chain: IBeaconChain;
   private network: INetwork;
   private reps: ReputationStore;
   private logger: ILogger;
 
-  public constructor(opts: SyncOptions,
-    {config, db, chain, network, reps, logger}:
-    { config: IBeaconConfig; db: IBeaconDb; chain: IBeaconChain; network: INetwork; reps: ReputationStore; logger: ILogger }) {
+  public constructor(opts: ISyncOptions, {config, db, chain, network, reps, logger}:
```
Comment (mpetrunic) [inline packages/lodestar/src/network/libp2p/syncRpc.ts]: partial, also crete interface for module to make it more readable

File: packages/lodestar/src/node/nodejs.ts [hunk 19]
```diff
@@ -43,8 +43,8 @@ export class BeaconNode {
   public reps: ReputationStore;
   private logger: ILogger;
 
-  public constructor(opts: Partial<IBeaconNodeOptions>, {config, logger}: {config: IBeaconConfig; logger: ILogger}) {
-
+  public constructor(opts: Partial<IBeaconNodeOptions>, {config, logger}:
+  {config: IBeaconConfig; logger?: ILogger}) {
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: this is some weird line break

File: packages/lodestar/src/node/nodejs.ts [hunk 20]
```diff
@@ -54,45 +54,67 @@ export class BeaconNode {
...
       config,
-      controller: new LevelDbController(this.conf.db, {
-        logger: this.logger,
-      }),
+      controller: new LevelDbController(
+        {
+          ...this.conf.db,
+          loggingLevel: this.conf.loggingOptions[Module.DATABASE],
+        },
+        {logger}),
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: this aint right, because you've passed logger so it will ovveride logging level and module. That said, controller initialization should happen inside BeaconDb and get logger from BeaconDb

File: packages/lodestar/src/node/nodejs.ts [hunk 21]
```diff
@@ -54,45 +54,67 @@ export class BeaconNode {
...
-    });
-    this.eth1 = new EthersEth1Notifier(this.conf.eth1, {
-      config,
-      opPool: this.opPool,
-      logger: this.logger
+      db: this.db,
+      chain: this.chain,
+      network: this.network,
+      reps: this.reps,
+      logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: remove

File: packages/lodestar/src/node/nodejs.ts [hunk 22]
```diff
@@ -54,45 +54,67 @@ export class BeaconNode {
...
+    this.network = new Libp2pNetwork({
+      ...this.conf.network,
+      loggingLevel: this.conf.loggingOptions[Module.NETWORK],
+    }, {
       config,
-      db: this.db,
-      eth1: this.eth1,
-      logger: this.logger
+      libp2p: libp2p,
+      logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: remove

File: packages/lodestar/src/node/nodejs.ts [hunk 23]
```diff
@@ -54,45 +54,67 @@ export class BeaconNode {
...
     });
+    this.eth1 = new EthersEth1Notifier(
+      {
+        ...this.conf.eth1,
+        loggingLevel: this.conf.loggingOptions[Module.ETH1],
+      }, {
+        config,
+        opPool: this.opPool,
+        logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: remove

File: packages/lodestar/src/node/nodejs.ts [hunk 24]
```diff
@@ -54,45 +54,67 @@ export class BeaconNode {
...
+      });
+    this.chain = new BeaconChain(
+      {
+        ...this.conf.chain,
+        loggingLevel: this.conf.loggingOptions[Module.CHAIN],
+      }, {
+        config,
+        db: this.db,
+        eth1: this.eth1,
+        logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: remove

File: packages/lodestar/src/node/nodejs.ts [hunk 25]
```diff
@@ -101,7 +123,7 @@ export class BeaconNode {
       network: this.network,
       reps: this.reps,
       rpc,
-      logger: this.logger,
+      logger,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/nodejs.ts]: remove

File: packages/lodestar/src/node/options.ts [hunk 26]
```diff
@@ -21,6 +23,8 @@ export interface IBeaconNodeOptions {
   opPool: IOpPoolOptions;
   sync: ISyncOptions;
   validator?: IValidatorOptions;
+  loggingOptions?: LogLevel[];
```
Comment (mpetrunic) [inline packages/lodestar/src/node/options.ts]: switch it back to Map<Module, LogLevel>

File: packages/lodestar/src/node/options.ts [hunk 27]
```diff
@@ -33,7 +37,19 @@ export const BeaconNodeOptions: IConfigurationModule = {
...
     OpPoolOptions,
     SyncOptions,
-    ValidatorOptions
+    ValidatorOptions,
+    {
+      name: "loggingOptions",
+      type: String,
+      configurable: true,
+      cli: {
+        flag: "loggingOptions",
```
Comment (mpetrunic) [inline packages/lodestar/src/node/options.ts]: should be called logLevels

File: packages/lodestar/src/sync/index.ts [hunk 28]
```diff
@@ -103,7 +104,7 @@ export class Sync extends EventEmitter {
       opPool: this.opPool,
       logger: this.logger,
     });
-    this.syncer.start();
+    await this.syncer.start();
```
Comment (mpetrunic) [inline packages/lodestar/src/sync/index.ts]: I'm pretty sure this isn't suppose to be awaited, node should not wait for chain to synchronize

File: packages/lodestar/src/util/parse.ts [hunk 29]
```diff
@@ -0,0 +1,19 @@
+/**
+ * @module util/parse
+ */
+
+import {LogLevel, Module} from "../logger/abstract";
+
+export function parseLoggingLevel(loggingLevelWithModule: string): LogLevel[]{
```
Comment (mpetrunic) [inline packages/lodestar/src/util/parse.ts]: parseLoggingLevels*

**Review by mpetrunic — CHANGES_REQUESTED** [review]
You forgot to add defaul log level to each module options

Some small naming and bugs, but it LGTM

File: packages/lodestar/src/cli/commands/eth1-private-network.ts (hunk 15 above)
Comment (sadiq1971) [reply packages/lodestar/src/cli/commands/eth1-private-network.ts]: i think yes.

**Review by mpetrunic — DISMISSED** [review]

**Review by GregTheGreek — DISMISSED** [review]

**Review by wemeetagain — CHANGES_REQUESTED** [review]
@mpetrunic suggested I add a few commits on top since @sadiq1971 will be busy the next few days

File: packages/lodestar/src/chain/chain.ts [hunk 30]
```diff
@@ -49,9 +49,8 @@ export class BeaconChain extends EventEmitter implements IBeaconChain {
   private logger: ILogger;
   private opts: IChainOptions;
   
-  public constructor(opts: Partial<IChainOptions>, {config, db, eth1, opPool, logger}: IBeaconChainModules) {
+  public constructor(opts: IChainOptions, {config, db, eth1, opPool, logger}: IBeaconChainModules) {
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: why did you remove partial?

Comment (wemeetagain) [reply packages/lodestar/src/chain/chain.ts]: resolved via chat. Saving adding Partial<Options> to all constructors as a future PR.

**Review by mpetrunic — CHANGES_REQUESTED** [review]
few points:
- first I think you missed the point, logLevel per module cannot be configured now aka we don't have `-logLevel chain=debug,network=info`
   - idea was to set custom processing function in BeaconLoggerOptions to parse that string
- second, I think you should create method in logger module to create logger from level and name, because if we ever wan't to change something we would have to modify  bunch of constructors throughout code

**Comment by wemeetagain** [comment]
>     * first I think you missed the point, logLevel per module cannot be configured now aka we don't have `-logLevel chain=debug,network=info`

Ah, I didn't put `cli` options in `src/node/loggerOptions.ts`!

>     * second, I think you should create method in logger module to create logger from level and name, because if we ever wan't to change something we would have to modify  bunch of constructors throughout code

The WinstonLogger constructor optionally takes an object of `{level: LogLevel, module: string}`, so we can configure them as needed.
My refactor removed the logger constructors from our core modules, and keeps the logger as an externally configured dependency to be injected. They're currently all instantiated in the `src/node/nodejs.ts`.

**Comment by mpetrunic** [comment]
> The WinstonLogger constructor optionally takes an object of `{level: LogLevel, module: string}`, so we can configure them as needed.
> My refactor removed the logger constructors from our core modules, and keeps the logger as an externally configured dependency to be injected. They're currently all instantiated in the `src/node/nodejs.ts`.

I mean't it would be better to add function 
```
createLogger(level, module) {
   return new WinstonLogger...
}
```
So it's easier to swap logger or change constructor.


Side note: How would be enable file based logging now that we split logger? Does that work with multiple loggers wrting single file?

**Comment by wemeetagain** [comment]
> So it's easier to swap logger or change constructor.

I added a `child` method to ILogger, which creates a child logger based on the parent's underlying logger instance, so we can create new loggers without messing with concrete logger constructors. Now the WinstonLoggers are created in cli and passed into the src/node and any child logger is created using `child`.

> Side note: How would be enable file based logging now that we split logger? Does that work with multiple loggers wrting single file?

Now this should work, no questions, since we're using the build in child functionality of Winston (and other logger libraries support this as well).

File: packages/lodestar/src/node/loggerOptions.ts [hunk 31]
```diff
@@ -0,0 +1,134 @@
...
+  description: "beacon chain logger options",
+  fields: [
+    {
+      name: "chain",
+      fields: [
+        {
+          name: "level",
+          description: "log level",
+          type: "string",
+          configurable: true,
```
Comment (mpetrunic) [inline packages/lodestar/src/node/loggerOptions.ts]: Those configs won't expose cli flags! it's missing [cli property](https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar/src/util/config.ts#L8).

That said naming is quite awkward. I would much prefer single flag `logLevel` with [custom processing function to parse input](https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar/src/util/config.ts#L7)

**Review by mpetrunic — CHANGES_REQUESTED** [review]

File: packages/lodestar/src/node/loggerOptions.ts (hunk 31 above)
Comment (wemeetagain) [reply packages/lodestar/src/node/loggerOptions.ts]: :+1: I'll add the custom processing

**Review by mpetrunic — COMMENTED** [review]
looks good, I will approve after you fix conflicts :)

**Review by mpetrunic — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]
